### PR TITLE
test(integration): username and password flow

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

test(integration): username and password flow

## Linear Issue Reference

* LOG-3022

## Testing

image

### CODE REVIEW

File: packages/integration-tests/src/logto-context.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    this.cookie = cookie;
+  }
+
+  public getCodeVerifier() {
+    return this.codeVerifier;
+  }
+
+  public getState() {
+    return this.state;
+  }
```
Comment (IceHe) [inline packages/integration-tests/src/logto-context.ts:59]: maybe make them public or use js property getter?
seems that there is no need to encapsulate these simple getters like java/kotlin?

Comment (xiaoyijun) [reply packages/integration-tests/src/logto-context.ts]: updated

File: packages/integration-tests/jest.setup.js [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+// eslint-disable-next-line unicorn/prefer-module
+const fetch = require('node-fetch');
+
+/* eslint-disable @silverhand/fp/no-mutation */
+global.crypto = new Crypto();
+global.fetch = fetch;
+global.TextDecoder = TextDecoder;
+global.TextEncoder = TextEncoder;
+/* eslint-enable @silverhand/fp/no-mutation */
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/jest.setup.js:14]: `ts` file will cause incompatible problems when use `node-fetch` and `text-encoder`

File: packages/integration-tests/src/logto-context.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  setData: (key: ContextDataKey, value: string) => void;
+  getData: (key: ContextDataKey) => string;
+};
+
+const createContextData = (): ContextData => {
+  const data: { [key in ContextDataKey]: string } = {
+    codeVerifier: '',
+    state: '',
+    interactionCookie: '',
+  };
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/src/logto-context.ts]: updated

File: packages/integration-tests/src/constants.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+export const logtoUrl = getEnv('LOGTO_URL');
+
+export const adminConsoleAppId = 'admin_console';
+
+export const redirectUri = `${logtoUrl}/console/callback`;
+
+export const authorizationEndpoint = `${logtoUrl}/oidc/auth`;
+
+export const tokenEndpoint = `${logtoUrl}/oidc/token`;
```
Comment (simeng-li) [inline packages/integration-tests/src/constants.ts:11]: get from oidc ?

File: packages/integration-tests/src/constants.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { getEnv } from '@silverhand/essentials';
+
+export const logtoUrl = getEnv('LOGTO_URL');
+
+export const adminConsoleAppId = 'admin_console';
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/src/constants.ts]: `@logto-schema` is not a public package.
https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1655265745355949

File: packages/integration-tests/src/logto-context.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  return {
+    account,
+    getCodeVerifier: () => getData('codeVerifier'),
+    getCodeChallenge: () => getData('codeChallenge'),
+    getState: () => getData('state'),
+    getAuthorizationCode: () => getData('authorizationCode'),
+    getInteractionCookie: () => getData('interactionCookie'),
+    getNextRedirectTo: () => getData('nextRedirectTo'),
+    getAuthorizationEndpoint: () => getData('authorizationEndpoint'),
+    getTokenEndpoint: () => getData('tokenEndpoint'),
```
Comment (simeng-li) [inline packages/integration-tests/src/logto-context.ts:109]: nit: not a big deal, but feels like export setData and getData could be good enough? as the input type is guarded by typescript already.

File: packages/integration-tests/tests/username-password-flow.test.ts [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+    setAuthorizationCode,
+    setNextRedirectTo,
+    setUpEndpoints,
+  } = createLogtoContext({
+    username: generatePassword(),
+    password: generateUsername(),
+  });
+
+  beforeAll(async () => {
+    await initInteraction();
```
Comment (simeng-li) [inline packages/integration-tests/tests/username-password-flow.test.ts:38]: nit: how about
```
logtoContext.init();
```

or make it as a class: LogtoContext, and new an instance at the very beginning. 
and properties can be accessed directly. 

logtoContext.codeChallenge
logtoContext.state

**Review by simeng-li — APPROVED** [review]
overall lgtm, thanks for the efforts.

File: packages/integration-tests/src/logto-context.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  getAuthorizationEndpoint: () => string;
+  getTokenEndpoint: () => string;
+  initInteraction: () => Promise<void>;
+  updateCookie: (cookie: string) => void;
+  setNextRedirectTo: (redirectTo: string) => void;
+  setAuthorizationCode: (authorizationCode: string) => void;
+  setUpEndpoints: (endpoints: { authorizationEndpoint: string; tokenEndpoint: string }) => void;
+};
+
+export const createLogtoContext = (account: Account): LogtoContext => {
```
Comment (IceHe) [inline packages/integration-tests/src/logto-context.ts:64]: maybe declaring a class is better than declaring a type and its creator

File: packages/integration-tests/tests/username-password-flow.test.ts [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+    const { redirectTo: completeAuthUrl } = consentResponse;
+
+    expect(completeAuthUrl).toBeTruthy();
+    expect(completeAuthUrl.startsWith(`${logtoUrl}/oidc/auth`)).toBeTruthy();
+
+    setNextRedirectTo(completeAuthUrl);
+  });
+
+  it('should get the authorization code from the callback uri when the auth process is completed', async () => {
```
Comment (IceHe) [inline packages/integration-tests/tests/username-password-flow.test.ts:158]: seems that: the authorization process should be completed after fetching tokens by authorization code?

**Review by IceHe — APPROVED** [review]
Overall LGTM

File: packages/integration-tests/tests/username-password-flow.test.ts [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+      codeChallenge: getCodeChallenge(),
+      state: getState(),
+    });
+
+    const response = await got(signInUri, {
+      followRedirect: false,
+    });
+
+    // Note: this will redirect to the ui sign in page
+    expect(response.statusCode).toBe(303);
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/tests/username-password-flow.test.ts:63]: good call, updated.

File: packages/integration-tests/tests/username-password-flow.test.ts (hunk 9 above)
Comment (xiaoyijun) [reply packages/integration-tests/tests/username-password-flow.test.ts:156]: No, the authorization process returns the authorization code.
the fetch token step comes after the process is completed.

File: packages/integration-tests/tests/username-password-flow.test.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+      },
+    });
+
+    expect(authCodeResponse).toHaveProperty('statusCode', 200);
+    const callbackUri = authCodeResponse.redirectUrls[0];
+    expect(callbackUri).toBeTruthy();
+
+    if (!callbackUri) {
+      throw new Error('No redirect uri');
+    }
```
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/tests/username-password-flow.test.ts:170]: The line `expect(callbackUri).toBeTruthy();` does not guard the `callbackUri`'s type, we handle the null case in this way.

[the remark this answers is not in the document]
Comment (IceHe) [reply packages/integration-tests/tests/username-password-flow.test.ts:152]: could be `expect(callbackUri).not.toBeNull();`?

File: packages/integration-tests/tests/username-password-flow.test.ts [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+
+    // Note: Redirect to consent page
+    expect(invokeAuthResponse).toHaveProperty('statusCode', 303);
+    expect(invokeAuthResponse.body).toBe(
+      'Redirecting to <a href="/sign-in/consent">/sign-in/consent</a>.'
+    );
+
+    const cookie = extractCookie(invokeAuthResponse);
+    expect(cookie).toBeTruthy();
+    expect(cookie.includes('_session.sig')).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/integration-tests/tests/username-password-flow.test.ts:112]: nit: more accurate: `expect(cookie).toBeTruthy();` → `expect(cookie).toBeDefined();`?

[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/tests/username-password-flow.test.ts:112]: use `Truthy` to guard empty string.

File: packages/integration-tests/tests/username-password-flow.test.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (xiaoyijun) [reply packages/integration-tests/tests/username-password-flow.test.ts:152]: ditto

File: packages/integration-tests/src/logto-context.ts (hunk 6 above)
Comment (xiaoyijun) [reply packages/integration-tests/src/logto-context.ts]: updated

**Pull request merged**